### PR TITLE
[IMP] open_academy: list coloring, calendar, search view T#59013

### PR DESCRIPTION
**Pull request by geraspace**
- Conditional colored sessions (dependent on duration of course)
- Calendar view
- Filter and group in search view

### CODE REVIEW

**Comment by geraspace** [comment]
Good morning @edy1192, here's a small update. Thank you!

File: open_academy/models/open_academy_course.py [hunk 1]
```diff
@@ -10,3 +10,21 @@ class OpenAcademyCourse(models.Model):
 
     responsible_id = fields.Many2one('res.users')
     session_ids = fields.One2many('open.academy.session', 'course_id')
+
+    _sql_constraints = [
+        ('name_description',
+         'CHECK(name != description)',
+         "The course title must be different to the description"),
+
```
Comment (edy1192) [inline open_academy/models/open_academy_course.py]: Remove blank line please.

File: open_academy/models/open_academy_course.py [hunk 2]
```diff
@@ -10,3 +10,21 @@ class OpenAcademyCourse(models.Model):
 
     responsible_id = fields.Many2one('res.users')
     session_ids = fields.One2many('open.academy.session', 'course_id')
+
+    _sql_constraints = [
+        ('name_description',
+         'CHECK(name != description)',
+         "The course title must be different to the description"),
```
Comment (edy1192) [inline open_academy/models/open_academy_course.py]:
```suggestion
         "The course title must be different to the description."),
```

File: open_academy/models/open_academy_course.py [hunk 3]
```diff
@@ -10,3 +10,21 @@ class OpenAcademyCourse(models.Model):
...
+
+        ('name_unique',
+         'UNIQUE(name)',
+         "The course title must be unique"),
```
Comment (edy1192) [inline open_academy/models/open_academy_course.py]:
```suggestion
         "The course title must be unique."),
```

File: open_academy/models/open_academy_course.py [hunk 4]
```diff
@@ -10,3 +10,21 @@ class OpenAcademyCourse(models.Model):
...
+    ]
+
+    @api.returns("self", lambda value: value.id)
+    def copy(self, default=None):
+        if default is None:
+            default = {}
+        if not default.get("name"):
+            default["name"] = "Copy of " + self.name
+        return super(OpenAcademyCourse, self).copy(default)
```
Comment (edy1192) [inline open_academy/models/open_academy_course.py]: Did you use precommit on this change?

Call the super method with parameters is no longer used, please run precommit and change this.

File: open_academy/models/open_academy_course.py [hunk 5]
```diff
@@ -10,3 +10,21 @@ class OpenAcademyCourse(models.Model):
...
+        ('name_unique',
+         'UNIQUE(name)',
+         "The course title must be unique"),
+    ]
+
+    @api.returns("self", lambda value: value.id)
+    def copy(self, default=None):
+        if default is None:
+            default = {}
+        if not default.get("name"):
```
Comment (edy1192) [inline open_academy/models/open_academy_course.py:27]: Why this validation?

File: open_academy/models/open_academy_session.py [hunk 6]
```diff
@@ -41,3 +42,9 @@ def _onchange_seats(self):
                     "message": "The number of attendees can not be greater than the seats available"
                 }
             }
+
+    @api.constrains("instructor_id", "attendee_ids")
+    def _check_instructor_in_attendee(self):
+        for record in self:
+            if record.instructor_id and record.instructor_id in record.attendee_ids:
+                raise ValidationError(_("A session's instructor can't be an attendee"))
```
Comment (edy1192) [inline open_academy/models/open_academy_session.py]: Why are you using the `_` method? this method is for translations. And you still haven't reached that part of the course 🤔

File: open_academy/models/open_academy_session.py [hunk 7]
```diff
@@ -41,3 +42,9 @@ def _onchange_seats(self):
                     "message": "The number of attendees can not be greater than the seats available"
                 }
             }
+
+    @api.constrains("instructor_id", "attendee_ids")
+    def _check_instructor_in_attendee(self):
+        for record in self:
+            if record.instructor_id and record.instructor_id in record.attendee_ids:
```
Comment (edy1192) [inline open_academy/models/open_academy_session.py]: You could improve this
```suggestion
        for record in self.filtered('instructor_id'):
            if record.instructor_id in record.attendee_ids:
```

**Review by edy1192 — CHANGES_REQUESTED** [review]

**Comment by geraspace** [comment]
@edy1192 it's ready to be checked. Thanks!

File: open_academy/models/open_academy_session.py [hunk 8]
```diff
@@ -31,13 +32,20 @@ def _onchange_seats(self):
...
                     "title": "Incorrect values",
-                    "message": "The number of attendees can not be greater than the seats available"
+                    "message": "The number of attendees can not be greater than the seats available."
                 }
             }
+
+    @api.constrains("instructor_id", "attendee_ids")
+    def _check_instructor_in_attendee(self):
+        for record in self.filtered('instructor_id'):
+            __import__('pdb').set_trace()
```
Comment (edy1192) [inline open_academy/models/open_academy_session.py]: Please remove this pdb.

**Review by edy1192 — CHANGES_REQUESTED** [review]

File: open_academy/models/open_academy_session.py (hunk 8 above)
Comment (geraspace) [reply open_academy/models/open_academy_session.py]: hahaha sorry @edy1192 I was just playing around with that and again forgot to remove it. Thanks!

**Review by edy1192 — APPROVED** [review]

**Pull request merged**